storage/storage_image.go

```diff
@@ -0,0 +1,430 @@
+	Created        time.Time         `json:"created-time,omitifempty"`
+	ID             string            `json:"id"`
+	BlobList       []string          `json:"blob-list,omitifempty"` // Ordered list of every blob the image has been told to handle
+	Layers         map[string]string `json:"layers,omitifempty"`    // Map from names of blobs that are layers to layer IDs
+	BlobData       map[string][]byte `json:"-"`                     // Map from names of blobs that aren't layers to contents, temporary
+	SignatureSizes []int             `json:"signature-sizes"`       // List of sizes of each signature slice
+}
+
+type storageLayerMetadata struct {
+	Compression  archive.Compression `json:"compression"`
+	ExpectedSize int64               `json:"expected-size"`
+}
+
+// newImageSource sets us up to read out an image, which we assume exists.
+func newImageSource(imageRef *storageReference) *storageImage {
+	ref := imageRef.reference
+	tag := ""
+	if imageRef.id != "" {
+		ref = imageRef.id
+		tag = imageRef.reference
+	}
+	img, err := imageRef.store.GetImage(ref)
+	if err != nil {
+		return nil
+	}
+	image := storageImage{
+		store:    imageRef.store,
+		imageRef: imageRef,
+		Tag:      tag,
+		Created:  time.Now(),
+		ID:       img.ID,
+		BlobList: []string{},
+		Layers:   make(map[string]string),
+		BlobData: make(map[string][]byte),
+	}
+	image.loadMetadata()
+	return &image
+}
+
+// newImageDestination sets us up to write a new image.
+func newImageDestination(imageRef *storageReference) *storageImage {
+	// We set the image's ID if the reference we got looked like one, since
+	// we take that as an indication that it's going to end up with the
+	// same contents as an image we already have.  If the reference looks
+	// more like a name, we don't know yet if it'll be exactly the same as,
+	// or be an updated version of, an image we might already have, so we
+	// have to err on the side of caution and create a new image which will
+	// be assigned the name as its tag.
+	image := storageImage{
+		store:    imageRef.store,
+		imageRef: imageRef,
+		Tag:      imageRef.reference,
+		Created:  time.Now(),
+		ID:       imageRef.id,
+		BlobList: []string{},
+		Layers:   make(map[string]string),
+		BlobData: make(map[string][]byte),
+	}
+	return &image
+}
+
+func newImage(imageRef *storageReference) *storageImage {
+	return newImageSource(imageRef)
+}
+
+func (s *storageImage) loadMetadata() {
+	if s.ID != "" {
+		image, err := s.store.GetImage(s.ID)
+		if image != nil && err == nil {
+			json.Unmarshal([]byte(image.Metadata), s)
+		}
+	}
+}
+
+func (s *storageImage) saveMetadata() {
+	if s.ID != "" {
+		if metadata, err := json.Marshal(s); len(metadata) != 0 && err == nil {
+			istore, err := s.store.GetImageStore()
+			if istore != nil && err == nil {
+				istore.SetMetadata(s.ID, string(metadata))
+			}
+		}
+	}
+}
+
+func (s *storageImage) Reference() types.ImageReference {
+	s.ID = s.imageRef.ID()
+	return s.imageRef
```

Comment:
Yeah, getting the storageReference to re-check for an image ID that matches its reference, in case the image is asked for the reference that was used to set it up, can probably be done in Commit(), since that's where an ID is assigned if there isn't already an image with a specified ID.